playground/config/__tests__/resolveConfig.spec.ts

```diff
@@ -0,0 +1,54 @@
+import { createLogger, normalizePath, resolveConfig } from 'vite'
+import { expect, it } from 'vitest'
+
+const normalizeString = (str: string) => {
+  const myRoot = normalizePath(process.cwd())
+  // eslint-disable-next-line no-control-regex
+  str = str.replace(/\x1b\[[\d;]+m/gi, '').replace(myRoot, '')
+  return str
+}
+
+it('resolveConfig with root path including both "#" and "?" should warn for both charchters ', async () => {
+  const logger = createLogger('info')
+  logger.warn = (str) => {
+    expect(normalizeString(str)).toEqual(
+      'The project root contains the "#" and "?" characters (/inc?ud#s), ' +
+        'which may not work when running Vite. Consider renaming the ' +
+        'directory to remove the "#" and "?" characters.',
+    )
+  }
+
+  await resolveConfig({ root: './inc?ud#s', customLogger: logger }, 'build')
+
+  expect.assertions(1)
+})
+
+it('resolveConfig with root path including "#" should warn accordingly ', async () => {
+  const logger = createLogger('info')
+  logger.warn = (str) => {
+    expect(normalizeString(str)).toEqual(
+      'The project root contains the "#" character (/includ#s), which may ' +
+        'not work when running Vite. Consider renaming the directory to ' +
+        'remove the "#" character.',
+    )
+  }
+
+  await resolveConfig({ root: './includ#s', customLogger: logger }, 'build')
+
+  expect.assertions(1)
+})
+
+it('resolveConfig with root path including "?" should warn accordingly ', async () => {
+  const logger = createLogger('info')
+  logger.warn = (str) => {
+    expect(normalizeString(str)).toEqual(
+      'The project root contains the "?" character (/inc?udes), which may ' +
```

Comment:
I think we may be introducing to many, too specific, tests here for this feature. I would be ok with a single one and to have a includes instead of equal test so we avoid having to maintain all this in sync if we later change the wording.

Reply:
Agree 👍 I moved this as a unit test with only a single.